### ISSUE TITLE
Centralize job signals

### ISSUE THREAD

**Issue by delijati** [QMatrixClient/libqmatrixclient#80]
All jobs e.g. `joinedRoom` should be fired when the job `succeeds` or more general `returns` not when started (When i do a refresh of the GUI uMatriks the room is not yet removed). Should centralize `result` && individual job `events`? See https://github.com/QMatrixClient/libqmatrixclient/blob/master/jobs/basejob.cpp#L255

I also need this for `leftRoom` signal (not yet pushed).

And we should also be able to show the error from a job aka "room does not exist"

**Comment by KitsuneRal** [comment]
I'm not sure I understand the question; but since `BaseJob` and the library in general are not quite well documented, let me accidentally elaborate here, and probably you will find the answer. I might put this to some document later on.

First off, a small bit of background on the library architecture. The library has two layers which interact in as small number of places as possible:
- lower-level transport layer - it's basically a wrapper around the client-server "RESTful" API (aka CS API) documented at Matrix.org. It consists of a `Connection` class, Jobs, and Events and some smaller supplementary data structures. Jobs basically reflect CS API calls; Events and other data structures form a data model. `Connection` is the carrier of all that, it enables network interaction. All this is almost a direct reflection of [The Spec](https://matrix.org/speculator/spec/HEAD/client_server/unstable.html), and I'm now working to make this layer mostly auto-generated from The Spec. An important thing here is that there's absolutely no processing of events within this API - they are just marshalled/unmarshalled to/from JSON, and that's it. Encryption (but not key management), when it arrives, will mostly be implemented in this layer.
- higher-level application layer - this consists of `Room`s, `User`s, a part of `Connection` (see more below on that) that deals with room dispatching (creation, joining etc.) and other not yet written classes, e.g., for connection management, a room directory, push rules, account data, key management etc. Things are much more structured and convenient here, instead of ids and string references like `sender` you can deal with objects and business-level containers like a room with its timeline, a user with the avatar, things like that. The underlying code of this layer is responsible for taking information from Events and making this structure out of it. Events still exist in this layer too, but there's some facilitating data model on top of them.

Now to specific things. When we speak about `BaseJob`, we speak about the transport layer. There are no rooms and users here. `BaseJob::result()` is a lower-level universal signal, you can connect to it if you need to know when the job (that is - an API call to the server) finished with whatever status. `success()` and `failure()` can be used to filter out only successful calls or only failing calls, respectively. Once the job has finished, you can call `error()` to check whether it finished successfully, and `errorString()` will give you a human readable description of the error, if there's one.

Application-level signals (like `joinedRoom()`) are emitted after Events have been taken from the transport layer into the application layer and processed there. They are specific for specific application-level objects. There's no plan, and in general it's not a good idea, to consolidate all application-level signals under one mediation class because that means that everybody should talk to this class with no apparent benefit from the majority of its interface; this class would become a huge registry of signals (there'd likely be about 100 signals just to deal with API calls completion) and be a major architectural bottleneck (with any change in the CS API we'd have to change this megaclass). Early on when the library has just been created, `Connection` usurped a part of application layer; hence we have `Connection::newRoom()`, `Connection::joinedRoom()` etc. in it. Ideally, there should be a `RoomManager` class that would have those signals - and `Connection` would get back to its transport layer origins.

As said above, I'm not sure if I answered your question, but here's how the library is organised, for what it's worth.

**Comment by delijati** [comment]
Ok i get now the problem ... I'm not really a QT GUI guy and the whole signal chaining is a bit confusing.
Anyway i found a solution to my problem by adding a new event `leftRoom`. Following my patch:

```patch
diff --git a/connection.cpp b/connection.cpp
index 56628a0..e5bf587 100644
--- a/connection.cpp
+++ b/connection.cpp
@@ -208,7 +208,10 @@ JoinRoomJob* Connection::joinRoom(const QString& roomAlias)
 
 void Connection::leaveRoom(Room* room)
 {
-    callApi<LeaveRoomJob>(room->id());
+    auto job = callApi<LeaveRoomJob>(room->id());
+    connect( job, &BaseJob::success, [=] () {
+        emit leftRoom(room);
+    });
 }
 
 RoomMessagesJob* Connection::getMessages(Room* room, const QString& from) const
diff --git a/connection.h b/connection.h
index f0b097f..c94f2c2 100644
--- a/connection.h
+++ b/connection.h
@@ -100,6 +100,7 @@ namespace QMatrixClient
             void syncDone();
             void newRoom(Room* room);
             void joinedRoom(Room* room);
+            void leftRoom(Room* room);
 
             void loginError(QString error);
             void networkError(size_t nextAttempt, int inMilliseconds);

```

Here my handler that i use in my model:
```cpp
...
void RoomListModel::setConnection(QMatrixClient::Connection* connection)
{
... 
   connect( connection, &QMatrixClient::Connection::leftRoom, this, &RoomListModel::removeRoom );
...
}

void RoomListModel::removeRoom(QMatrixClient::Room* room)
{
    int position = m_rooms.indexOf(room);
    beginRemoveRows(QModelIndex(), position, position);
    m_rooms.removeAt(position);
    endRemoveRows();
}
```

**Comment by KitsuneRal** [comment]
Well, with the current `Connection` code, this addition makes a perfect sense! I'll be glad to merge a PR with this.

**Comment by KitsuneRal** [comment]
Closed by #81
